### DISCUSSION TITLE
lastComboMove/lastComboActionID Clarification

### DISCUSSION THREAD

**Discussion by vitharr137** [daemitus/XIVComboPlugin#338] [3 - Melee DPS]
I've been digging through the Struct and attempting to find where the parameters for lastComboActionID are derived from, to try to build the potential to track any previously used actionID rather than just weaponskill (and oddly spell? it seems to work with BLM as well) 

Is there any way to return the previously used Ability?

**Comment by MKhayle** [comment]
I was thinking about using it for the All-In-One Generation Legacies features but there's almost always an alternative to find out this info like that https://github.com/MKhayle/XIVComboExpanded/pull/310#issuecomment-2226291887 

And the final implementation ended up being https://github.com/MKhayle/XIVComboExpanded/commit/69c85e47b68074b1cb017def0913de3f728783ea

**Reply by MKhayle** [reply]
So I don't know (right now) any way to do this, but my thought about it is that we shouldn't need it in general

**Reply by vitharr137** [reply]
I've been able to use your link's similar logic in almost all cases as an alternate, even during mudra, but my main use for it still eludes me: tracking for NIN if the last step of a ninjutsu was successful. I've been tinkering with while conditions to trap the mudra input, but I cannot find a reliable way to break the loop without potentially using a variation of 'lastActionID' to confirm that, say, Raiton, was the casted action. I'd left off last year testing it unsuccessfully and it's been on my mind recently again

**Reply by MKhayle** [reply]
oh, I see what you mean. 
you may try to catch the button pressing queue and cache the last inputs, but honestly? that's kinda hazardous, chaotic, and probably unreliable.

**Reply by vitharr137** [reply]
Very XD and the lumina sheets scare me too much. The action timeline addon seems to use them for tracking oGCD inputs to offset them, but I'm mostly illiterate to how they're tracking it in code. There's a few promising offsets in the FFXIVClientStructs but I haven't found a successful way to draw them into methods yet:

```
    [FieldOffset(104)]
    public bool ActionQueued;

    [FieldOffset(108)]
    public ActionType QueuedActionType;

    [FieldOffset(112)]
    public uint QueuedActionId;
```
I'll keep digging to see if I can find anything, as I think it would unlock several useful tools overall, and I'll report back if I find anything, but if someone beats me to it that'd be nifty

**Reply by MKhayle** [reply]
maybe @Grammernatzi could help you there because I have no idea how to track this exactly

**Reply by kaedys** [reply]
So, one thing we definitely want to make sure we don't break here is handling actual _combos_ while other abilities have been used in between.  For example, on Viper, you can be mid combo and then use Dreadwinder + the Coils, or the entire Reawakened combo, and that doesn't break the underlying basic combo (and also doesn't change the lastComboID value).  This is especially important for 1-2-3 combos on jobs where there are not already compressed onto a single button, since we use lastComboID to determine what the next skill in the combo is (example, Reaper's 1-2-3 is not interrupted by using Gibbet/Gallows.  Machinist 1-2-3 is not interrupted by an Overcharge sequence, or by the Tool skills).

**Reply by vitharr137** [reply]
For sure. Honestly it seems best to separate it entirely from the notion of combo , since they aren't really tied to anything specific like the timer or a previous 'combo weaponskill'. Something alone, like lastActionID itself would be a better naming convention I'd imagine, while leaving the others intact as they are. Ideally, all it would do is return the previously used actionID, regardless of type, almost like OriginalHook does with upgraded actions

**Reply by kaedys** [reply]
I think it might actually be better to split it even further, into `lastActionID` and `lastGCDID`.  That gives us 3 levels of tracking for it, all of which I've found situations where they could be useful (though thus far, all of them having ways around it).

**Reply by vitharr137** [reply]
this code snippet: `OriginalHook(ActionManager.Instance()->QueuedActionId)` does return Ability ActionIDs for the split second they are in process which may be enough to break the while loop I'm working on but now that I know how to call some of those directly, I'll keep looking into how to get a more permanent (more than 400ms or so) result

**Reply by vitharr137** [reply]
I don't know how to do it, but I believe if there was a way to write the result of `ActionManager.Instance()->QueuedActionId` to a variable when it's not null, that would give a stable-ish way to see what the last used Ability was. It does not work for Weaponskills, it has to be something able to be queued, which seems to only be Ability OGCDs

**Reply by kaedys** [reply]
Does QueuedActionId also return results when the action isn't actually _queued_, but just cast when already off GCD or not otherwise animation locked?

**Reply by vitharr137** [reply]
Sadly it doesn't seem to, no. I'll keep looking

Interesting odd note: QueuedActionID knows what the original actionID being replaced by the plugin is that was being pressed. I use a few inaccessible actionIDs in the testing classes I work with, and was confused about the odd icons I was seeing get returned by the QueuedActionID icon until I realized it was giving me back Anatman's old icon rather than what it 'looked like' I was pressing via the IconReplacer

**Reply by vitharr137** [reply]
The only other lead I have at the moment is another addon named RotationSolver that appears to be using the Lumina sheets directly in this code:
```
using Action = Lumina.Excel.GeneratedSheets.Action;
    internal static unsafe void AddActionRec(Action act)
    {
        if (!Player.Available) return;

        var id = (ActionID)act.RowId;

        //Record
        switch (act.GetActionCate())
        {
            case ActionCate.Spell:
            case ActionCate.Weaponskill:
                LastAction = LastGCD = id;
                break;
            case ActionCate.Ability:
                LastAction = LastAbility = id;
                break;
            default:
                return;
        }

        if (_actions.Count >= QUEUECAPACITY)
        {
            _actions.Dequeue();
        }

        _timeLastActionUsed = DateTime.Now;
        _actions.Enqueue(new ActionRec(_timeLastActionUsed, act));
    }
```

And ActionTimeline itself leans on that original addon for the ActionCate definitions in their lines in TimelineManager.cs for searching the Lumina sheets:
```
using FFXIVClientStructs.FFXIV.Client.Game;
using Lumina.Excel.GeneratedSheets;
using RotationSolver.Basic.Data;
using Action = Lumina.Excel.GeneratedSheets.Action;
using Status = Lumina.Excel.GeneratedSheets.Status;
    public unsafe float GCD
    {
        get
        {
            var cdGrp = ActionManager.Instance()->GetRecastGroupDetail(GCDCooldownGroup - 1);
            return cdGrp->Total;
        }
    }

    private static TimelineItemType GetActionType(uint actionId, ActionType type)
    {
        switch (type)
        {
            case ActionType.Action:
                var action = Svc.Data.GetExcelSheet<Action>()?.GetRow(actionId);
                if (action == null) break;

                if (actionId == 3) return TimelineItemType.OGCD; //Sprint

                var actionCate = (ActionCate)(action.ActionCategory.Value?.RowId ?? 0);

                var isRealGcd = action.CooldownGroup == GCDCooldownGroup || action.AdditionalCooldownGroup == GCDCooldownGroup;
                return actionCate == ActionCate.AutoAttack
                    ? TimelineItemType.AutoAttack
                    : !isRealGcd && actionCate == ActionCate.Ability ? TimelineItemType.OGCD
                    : TimelineItemType.GCD;

            case ActionType.Item:
                var item = Svc.Data.GetExcelSheet<Item>()?.GetRow(actionId);
                return item?.CastTimes > 0 ? TimelineItemType.GCD : TimelineItemType.OGCD;
        }

        return TimelineItemType.GCD;
    }
```

**Reply by vitharr137** [reply]
DeathRecap's CombatEventCapture.cs also seems to use the Lumina sheets only for their actionID parsing to capture ability-based heals and damage timelines

**Comment by Grammernatzi** [comment]
So I feel if you wanted to detect the last move used no matter what, you'd want to look at how the ActionTimeline plugin (or whatever it is) does it. I haven't bothered with this yet but I can see how it would be really useful. I ended up using something similar from another plugin I forgot to track whether actions were grayed out so that things played nicer when leveling on alts (and it also ended up being useful for tracking when a lot of actions could be used).

**Reply by vitharr137** [reply]
I'm still on the hunt for lastActionID but what you referenced I believe I located in the IconReplacer.cs section of your fork:
```
    /// <summary>
    /// Gets bool determining if action is greyed out or not.
    /// </summary>
    /// <param name="actionID">Action ID.</param>
    /// <param name="targetID">Target ID.</param>
    /// <returns>A bool value of whether the action can be used or not.</returns>
    internal unsafe bool CanUseAction(uint actionID, uint targetID = 0xE000_0000)
    {
        return clientStructActionManager->GetActionStatus(ActionType.Action, actionID, targetID, false, true) == 0;
    }
```
Then adding that to CustomCombo.cs via:
```
    /// <summary>
    /// Gets bool determining if action is greyed out or not.
    /// </summary>
    /// <param name="actionID">Action ID.</param>
    /// <returns>A bool value of whether the action can be used or not.</returns>
    protected static bool CanUseAction(uint actionID) => Service.IconReplacer.CanUseAction(actionID);
```

Did allow me to get a step closer to one of my goals which was make a running list of BLU filler spells without having to untangle the nightmare that is their spellbook array using this type of thing. I am hoping to turn this into a larger array but am still in the testing phase:
```
return CanUseAction(BLU.Reflux) ? BLU.Reflux : BLU.SonicBoom
```
As an example, which is awesome, thank you. I'm still looking through to find how lastComboAction works at all, but I believe it's targetting the
```
[FieldOffset(96)]
public ComboDetail Combo;
````

section, and sending the 0x4 to indicate it's looking for the Action to return given 0 is Timer and 4 is Action in the offsets, but how it's calling it in PluginAddressResolver is not something I am familiar with yet, given I don't know how `public IntPtr ComboTimer { get; private set; }` works at all. If I can figure that out, I'm hoping to be able to do some sort of call to see if I can drag out the offsets of these, if they're even available:
```
    [FieldOffset(272)]
    public ushort LastUsedActionSequence;

    [FieldOffset(274)]
    public ushort LastHandledActionSequence;
```

**Reply by kaedys** [reply]
That `CanUseAction` is absolutely something we need, assuming it's returning whether actions that have pre-requisites, but where those pre-reqs are _not_ buffs (ex. the Dreadwinder combo enabling the Coils; Iaijutsu, when used with the Tsubame Ready buff, _consumes_ the buff and then also enables Tsubame).  I've been trying to find a way to put Tsubame on the main action keys for SAM, because I'm _terrible_ at remembering to use it now that it's not its own cooldown (and is exclusively predicated on having used Meikyo recently, and is _only_ available immediately following the next Midare or Tenka, use it or lose it), but there's no buff and it's not counted as a `lastComboAction`.  I'll try it with that `CanUseAction`.

**Reply by vitharr137** [reply]
It also seems possible that it can, as stated by Grammernatzi, replace all the level checks for a prerequisite. I'm testing a few things to see if it works as a replacement for buffs, but it definitely has lots of potentials for more dynamic, less hardcoded checks

**Reply by kaedys** [reply]
That would certainly make handling patch and expansion changes easier.

**Reply by kaedys** [reply]
Had to also add the definition and instantiation of the `clientStructActionManager`.  Sadly, it does not seem to trigger from at least Tsubame being usable.

**Reply by Grammernatzi** [reply]
I just want to point out that CanUseAction does have a downside: it only checks if actions are grayed out, regardless of context. The issue with this is that casting grays out actions, so it can be kind of messy with casters. It works pretty well otherwise, though.

**Reply by kaedys** [reply]
Nevermind! So, it turns out you need to use the transformed ID, not the base one.  You can't check `CanUseAction(Tsubame)`, you need to check `CanUseAction(Kaeshi: Setsugekka)` instead.  Which is...an annoying thing, and rather starkly limits it's utility, unfortunately.

Edit: as an example, instead of simply checking if `Tsubame` is usable, you have to check for all 4 possible abilities that can go on that button (Kaeshi Setsugekka, Kaeshi Goken, Tendo Kaeshi Setsugekka, and Tendo Kaeshi Goken)

**Reply by kaedys** [reply]
Haha, nope, nevermind, found a way around that:
```
            if (CanUseAction(OriginalHook(SAM.TsubameGaeshi)))
                return OriginalHook(SAM.TsubameGaeshi);
```
This works exactly how I was hoping it would.

**Reply by kaedys** [reply]
And feature complete and tested.  Now I _can't_ forget to use Tsubame, because it replaces all of my combo buttons when available.  @MKhayle this is lit.

**Reply by MKhayle** [reply]
I'll add this with the rework, handy indeed
